### DISCUSSION TITLE
Importing DX7II / TX802 sysex patches?

### DISCUSSION THREAD

**Discussion by solyarisoftware** [asb2m10/dexed#405] [Ideas]
Hi,

Many years ago (around 1993-1996), I programmed original patches for my DX7IIFD, focusing on ambient and wild sounds, using the fantastic but now discontinued Galaxy editor. Those patches are still available on https://solyaris.altervista.org/dx7.html .

Recently, watching some youtube videos comparison, I've come to appreciate the sound of Dexed, especially when compared to FM8 and Arturia products. 

I'm interested in importing and running my original patches on Dexed. However, from what I understand, Dexed only allows the import and editing of original DX7 sysex patches. Is that correct?

As you may be aware, the DX7II performance patch is akin to having the layers of two DX7s along with a performance patch on top. My question is: how can I reproduce a DX7II program patch using Dexed? If this isn't possible with the current version of Dexed, could a potential solution be to split my DX7II sysex file into two DX7 sysex files and run two instances of Dexed?

Additionally, do you know of any Windows OS DX* librarian (preferably free) that could assist in splitting my sysex DX7II file?

By the way, will Dexed support sysex import for the entire DX family?

Thanks,
Giorgio

**Comment by BobanSpasic** [comment]
Hi @solyarisoftware,

you got a wrong idea about the internal construction of the DX7II files. You do not need any file-split, but still - DX7II has much more parameters than Dexed. It means - you can't reproduce DX7II performances, neither voices on Dexed.
A basic DX7 voice-parameter set is called VCED (and this is what you get with Dexed). DX7II has additional parameters per voice - ACED. On the top of that, you have performance with additional parameters PCED.
So, theoretically, you could use the VCED-part from your performances in Dexed, but these aren't 1:1 compatible, a small conversion is needed. You will lose your controller settings (mod wheel, breath controller etc) and the panning effects.

**Comment by solyarisoftware** [comment]
Hi Boban,

I didn't mention any DX7II internals construction, so I don't understand your initial statement. I wrote, _"The DX7II performance patch is akin to having the layers of two DX7s along with a performance patch on top"_ which generally means exactly what you stated: a DX7II PROGRAM is made up of 2 (VCED+ACED) + 1 PCED on top.

For me, that means, at a high level: "having 2 DX7s"! Sure, to be precise, the ACED and PCED add some controllers and pan (not too good, honestly), but these are really "minor" points, at least for me, because in practice, what's interesting in the DX7II is just having 2 layered DX7s. As for me, the TX802 is like having 8 layered DX7s. I hope you understand the sound designer & musician aprroximated perspective. 

So, back to Dexed, I'm aware (because I tested it yesterday) that Dexed manages only DX7 (mark I) sysex files. That's why a possible workaround is to SPLIT (sorry for insisting) the SYSEX DX7II PERFORMANCE BANK (the full dump sysex file like the [alienazy.syx](http://solyaris.altervista.org/dx7iipatches/alienazi.zip) people can download from my website) into 2 DX7 VOICE BANKS. Maybe using [DXConvert](https://dxconvert.martintarenskeen.nl/). I have to tray using it for the scope.

In that way, I patiently remade a performance X running a number of Dexed instances, each one running a voice component.

Do you have a smarter solution?

**Comment by BobanSpasic** [comment]
Hi Giorgio.

I can't say I have smarter solution, but I can write the app you need. I
just need the details

Giorgio Robino ***@***.***> schrieb am Di., 23. Jän. 2024,
10:10:

**Comment by solyarisoftware** [comment]
Thanks. I take a look of your open source projects; great work. 

let me try to expose better a possible general "need":
when concerning DX7II or a TX802 it could be interesting for any people have hw synths or just the sysex bank dump files (I call "programs") to play original patches on a virtual emulator (VST) environment.

So far, all VST emulator available (and here maybe dexed is the one sound better, in my quick listening through comparisons on youtube), just import a basic DX7 patch (VCED+ACED). 

Now, premising we want to play a DX7II program patch of mine through dexed, one solution is to run in parallel two dexed instance in parallel (that's easy in any DAW). Suppose my original program is name MY_PROGRAM, and its' composed by 2 VOICES: VOICE_A + VOICE_B, I would like to 
- export the program in two files VOICE_A.syx and VOICE_B.syx  
- load VOICE_A.syx in dexed VST first instance run-time inside a DAW  
- load VOICE_B.syx in dexed VST second instance run-time  
- play the layers and trickery with DAW / external MIDI CC controllers.

So the feature needed is to transform a DX7II program bank (the entire full bank composed by 1 32 programs bank+2 32 voices banks +...) in 2 DX7 voice banks. 

 BTW, I guess the same could be done for TX802. I guess there are many people own a DX7II, whereas the TX802 is much less spread.   
  
 Let me know if the need is clear.

**Comment by BobanSpasic** [comment]
Just don't take my words for rude, it is just a limit set up by my English knowledge.
1. Basic DX7 patch is just a VCED, no ACED involved. ACED is DX7II and TX802 only
2. We need to export voice banks in VMEM format (32 patches). Most of the emulators does not handle VCEDs (one-voice file)
3. If your performance uses one voice from Bank A, and one voice from Bank B - your setup described above will work OK. But, if you use two voices from the same bank, you need to load the same bank into both Dexed instances. Not a real problem, but it can't be done auto-magically
4. Some parameters from DX7II must be converted to DX7, and this does not work always as expected. Manual tuning of the voices is necessary. The two problematic parameters are AMS and PEGR. AMS for DX7II has a range from 0 to 7, and it is 0 to 3 for DX7. PEGR for DX7 is a parameter for 8 octaves range, and for DX7II it is a parameter over a selectable range (1/2 octave, 1 octave, 4 octaves, 8 octaves). So, the resolution of these two parameters is much higher for DX7II. The approximation to the nearest DX7 value won't make us happy in some cases.
5. Will a command-line tool do a job for you (I could finish it in 1-2 days), or you are more comfortable with a graphical interface (this would take some time, but it would be probably more convenient to use)?
6. As for the TX802 - not really a problem, but I didn't find so much TX802-specific files on the internet. I even had troubles finding enough files to get them analyzed, so that I can write a reader for these. Someone with a TX802 at the hand could probably help a bit here, so that I can get all the parameters from sysex dumps decoded.

**Comment by solyarisoftware** [comment]
Hi Boban,

Regarding your first sentence, don't worry; I also acknowledge that there is some cultural "independence mismatch" across people and languages all around the world. I understand your point of view, which is rightfully very technical in your dissections, whereas I am now a bit distant from the construction of synth programming patches (I programmed my DX7II sounds in the '90s!). That said, let me quote your great points:

---
> 1. Basic DX7 patch is just a VCED, no ACED involved. ACED is DX7II and TX802 only

That's now clear! Managing VCED without ACED is inevitable and it's an acceptable limit, in my opinion. 

---

> 2. We need to export voice banks in VMEM format (32 patches). Most of the emulators does not handle VCEDs (one-voice file)

OK

---

> 3. If your performance uses one voice from Bank A, and one voice from Bank B - your setup described above will work OK. But, if you use two voices from the same bank, you need to load the same bank into both Dexed instances. Not a real problem, but it can't be done auto-magically

Agreed. You are right. Maybe my programs often are build with voices from the same banks, if I well remember, but yes a program could be built with a voice from bank A or from bank B, so we do need the association:

```
program_name = voice_1_name (bank_name/voice1_name) + voice_2_name (bank_name/voice2_name)
```

E.g. 

```
PitecantropusErectus = A:Nuvole 3  + B:Nuvole 2
```
See the appendix here below for more details.

---
 
> 4. Some parameters from DX7II must be converted to DX7, and this does not work always as expected. Manual tuning of the voices is necessary. The two problematic parameters are AMS and PEGR. AMS for DX7II has a range from 0 to 7, and it is 0 to 3 for DX7. PEGR for DX7 is a parameter for 8 octaves range, and for DX7II it is a parameter over a selectable range (1/2 octave, 1 octave, 4 octaves, 8 octaves). So, the resolution of these two parameters is much higher for DX7II. The approximation to the nearest DX7 value won't make us happy in some cases.

That's IMPORTANT (and a bad news)! Thanks for sharing this, I admit I was not aware. 
For PEGR maybe I leved defaults but maybe no, because often I used randomized values using Opcode Galaxy II randomizer feature (a great discontinued multi-synt librarian editor on the Mac in 90s')  so you this also maybe explains why for my ears DX/II patches sound better than DX7... ?

---

> 5. Will a command-line tool do a job for you (I could finish it in 1-2 days), or you are more comfortable with a graphical interface (this would take some time, but it would be probably more convenient to use)?

Good point. For me personally (I'm a developer/linux/vim guy) command line is always better ah ah :-) but fro many people maybe the GUI/webGUI could be best accessible. Maybe initially a CLI program could be more than enough, and maybe the cli program to be integrated in a successive GUI... Which programming language you would use for the scope?

So maybe the tool could take in input the sysex file and generate in output 2 voice banks + a text file with the program/voices association? Something like:  

```
$ dx7ii_split_tool --input alienazy.sys
# generates in output 3 files:
#
# output File 1: alienazy.csv
# a text/CSV file to map program to voices? E.g. a CSV file like:
# Program_name; Bank A Voice Name; Bank B Voice Name 
# PitecantropusErectus; Nuvole 3; Nuvole 2
# Antropos Spaces; ...; ... 
#  
# output File 2: alienazy.voices_a.sys
# output File 3: alienazy.voices_b.sys
``` 

BTW, And/or maybe author of great [DX7II-Librarian](https://github.com/bladeSk/DX7II-Librarian) could be interested to integrate the feature in his great web GUI DX* librarian? 

---

> 6. As for the TX802 - not really a problem, but I didn't find so much TX802-specific files on the internet. I even had troubles finding enough files to get them analyzed, so that I can write a reader for these. Someone with a TX802 at the hand could probably help a bit here, so that I can get all the parameters from sysex dumps decoded.

Yes. I never had in my hands a TX802 unfortunately. The reason why you can find few TX802 patches is pretty simple: it depends on statistic of selling. DX7 synths are very common in 90s, DX7II much less spread, TX802 was very expensive and few rich people own one (at the times) :-) 


---


Here below the MIDI Quest DEMO editor screenshots that show how my _Alienazi_ DX7II program is composed in terms of voices. BTW the Midi quest program doesn't show voices on bank B (from 32 to 64)  ?? I don't know why. My Programs  refers to the voice on bank B too. What a mess. I never used Midi quest and honestly I do not trust this old demo program version I just installed few weeks ago.   

Lats but not least pretty all my programs are made as "DUAL" that just means that the two component voices are layered/overlapped (see program called _PitecantropusErectus_. 
   
![alienazi-midi-quest](https://github.com/asb2m10/dexed/assets/4106440/ffb6ca1b-14bd-4785-8e42-430d8f6cf1b5)

**Comment by BobanSpasic** [comment]
I am already writing the converter. I'll get back here with a test version in a couple of days. My RL job is a bitch at the moment - I am working 10–11 hours per day till the end of the month.
Do you still have a DX7II, so that you can compare the original with the converted files?

**Comment by solyarisoftware** [comment]
A bit the same for me. I fully understand RL's point and that's why I stop to make music for more than 15 years. As an old software engineer myself, please take into consideration reducing your daily hourly rate drastically to save your health. Please postpone this activity!

Yes, I own my DX7II, but it's been buried in the back of a closet for twenty years, and if I were to take it out for comparisons and playing patches, I'd surely be tempted to sample my old programs to build a sample library (maybe for Decent Sampler?) thus creating a new problem to myself ;-) but I have to get my hands on it, sooner or later. Right.

By the way, concerning the converter CLI tool, it comes to mind that it could be useful to warn the user on stdout prints regarding AMS and PEGR parameter mismatches/remaps, maybe with a --verbose, or maybe writing in the proposed CSV/report file the  mapping (I know, that concerns 2 voices per program).

I would also appreciate comments from other readers of this thread.

Thanks & slow down where possible :-)
Giorgio

**Comment by BobanSpasic** [comment]
https://github.com/BobanSpasic/MK2toMK1
Just a Windows version at the moment. Do you need Linux version?

**Reply by SamuelSHrk** [reply]
does this work with TX81Z/DX11/DX100,etc. SYX patches as well?

**Reply by BobanSpasic** [reply]
No way a DX7 can play DX11 voices, if that is your question at all.
DX11 and the other 4-OP DX synths are using 8 different waves for the oscillators. 
DX7 and other 6-OP DX synths are sinus-wave only. 
You can theoretically translate just these DX11 voices that are using sine-waves.
Take a look at the DXConvert (mentioned somewhere at the beginning of this thread) - it can do some translation of 4-OP to 6-OP by ignoring the wave differences. If you have luck, you'll get some usable results.

**Comment by solyarisoftware** [comment]
Great! 
I deepen and try to run your solution in the we.
 
In general I tend to program just on Linux (I use evey day WSL on a Windows). That's could be interesting also for portability on Mac.  But doesn't your Pascal CLI program be independent from the OS? 
BTW, I developed in TurboPascal in 80's ... but it's decades I don't see any pascal code. It's Free Pascal ok (for Linux)?

Thanks for your time!

**Comment by BobanSpasic** [comment]
It is OS-independent, but I need to fire up my Linux PC and my VM with OSX in order to compile it for these platforms. I can't cross-compile from Windows (my main development platform).
I've used Delphi in the past, but FreePascal/Lazarus evolved so much in the last couple of years that I do not need Delphi anymore. Lazarus visual components are cross-platform (Windows, GTK, QT, Carbon, Cocoa, some web-platforms etc.).

**Comment by solyarisoftware** [comment]
I see. Thanks! I'll try compile myself soon using free Pascal compiler I just installed on my Linux Ubuntu (WSL on Windows).

**Reply by BobanSpasic** [reply]
I've added the Linux binary to the downloads. It depends on libc6.
To compile it yourself, you'll need the commandline parameters for the compiler. I'll provide a shell script soon.

**Comment by BobanSpasic** [comment]
Any feedback? Did it work for you?

**Reply by Casuallynoted** [reply]
Hello! I'm currently trying to use this but the -o argument doesn't output a file. -i does give me info on the file however.

**Reply by BobanSpasic** [reply]
Hello. Could you post the output of the info from -i here?

**Reply by Casuallynoted** [reply]
Absolutely! This is what I'm getting:

<img width="267" alt="Screenshot 2025-06-30 at 12 30 17 AM" src="https://github.com/user-attachments/assets/6a2eb107-e1d6-4e81-8ced-ed4afff7be2f" />

**Reply by BobanSpasic** [reply]
It is a performance file, not a voice bank. A performance file is an additional file to a voice bank, and it doesn't contain the voice data itself.

**Reply by Casuallynoted** [reply]
Oh! Is there any way to know what voice bank it was meant to be used with? And is there any way it might be loaded in a VST like Dexed? I'm still pretty new to performance files so this is really fascinating!

**Reply by BobanSpasic** [reply]
Unfortunately, no. There are no references to the voice bank in the performance file.
Also, it is not usable with Dexed in any automatic way. One DX7II performance would span over two Dexed engine (split, duet etc. - using two voices at the same time).

**Comment by probonopd** [comment]
I've been working on __FMRack__, an open source synthesizer engine roughly modeled after the TX816 that uses multiple Synth_Dexed engine instances (currently up to 128).

Given enough time and dedication, I believe it might be possible to support DX7II performances natively. For now it supports MiniDexed performances, a format for which @BobanSpasic has written converters. 

In this demo you hear 3 FMRack instances playing using slightly changed TX816 performances (be sure to unmute):

https://github.com/user-attachments/assets/dbdc602b-e466-4333-9137-e159499dfb5e

The FMRack engine is C++. There are pure command line versions (e.g., to run on an embedded device like a Raspberry Pi) and there are GUI versions (standalone and VST3).

Right now it is a functional proof-of-concept. Would anyone like to collaborate? Ideally you'd have C++ and possibly JUCE skills.

**Comment by maxcomperatore** [comment]
For anyone looking for a modern way to manage, inspect, and split DX7 and DX7II SysEx banks without having to run 32-bit Windows XP utilities or DOS-era software:

We built [bipluk](https://bipluk.com) specifically around this problem. It's a zero-install Web MIDI SysEx librarian running directly in Chrome:

* **DX7 & DX7II 32-Voice Parsing:** When you drop or capture a full 4,096-byte Yamaha bulk dump, it parses the byte-stream with bitmask unpacking, extracting ASCII patch names directly from the VCED/ACED parameter blocks.
* **Single-Voice & Bank Split:** You can audition individual patches, export clean single-voice .syx files (VMEM/VCED), and organize banks in your browser vault.
* **Direct Hardware Handshake:** It communicates directly with physical DX7/DX7II/TX802 hardware over Web MIDI with automated 31,250 baud byte-pacing (60ms chunk delay) to avoid buffer overflow errors on vintage 8-bit CPUs.

Hope this helps anyone struggling with vintage Yamaha FM dump formats!

**Reply by probonopd** [reply]
Can you point me to the source code please? Is [this](https://github.com/maxcomperatore/hardware-stream-bridge) it?

Looks like [this](https://github.com/maxcomperatore/hardware-stream-bridge/blob/main/sysex_adapters/YamahaDX7II.py) is the relevant source code that would need to be ported over.

**Reply by maxcomperatore** [reply]
Yes, exactly! That file (`sysex_adapters/YamahaDX7II.py`) defines the exact 128-byte packed bitfield layout (`OPERATOR_PACKED` and `VOICE_PACKED`).

Notice how the bitfield definitions combine parameters into single bytes:
- `osc_mode` (1 bit) + `freq_coarse` (5 bits)
- `detune` (4 bits) + `rate_scale` (3 bits)
- `amp_mod_sensivity` (2 bits) + `key_vel_sensivity` (3 bits)

In C++, porting that over to FMRack is very clean using `#pragma pack(push, 1)` structs with matching bitfield widths, or standard bitwise masks (`(byte >> offset) & mask`). 

In Bipluk's browser client, we unpack those exact bitmasks in JavaScript to render the patch names and FM operator parameters in real time as the bulk dump streams in.

If you need any specific DX7II/TX802 test dumps or parameter tables to test with FMRack, let me know!
